# **Level 6 Data Science / Software Engineering**

# Topic 3 &mdash; Introduction to Pandas


## The Problem: Handling Complex Data

In data science, we often encounter large, complex datasets that can be challenging to work with. These datasets may come from various sources and in different formats, such as:

- CSV files
- Excel spreadsheets
- SQL databases
- JSON files
- Web APIs

Working with this data efficiently and effectively presents several challenges:

1. **Data cleaning**: Real-world data is often messy, containing missing values, duplicates, or inconsistencies.
2. **Data manipulation**: Reshaping, merging, and transforming data to suit our analysis needs.
3. **Data analysis**: Performing calculations, aggregations, and statistical operations on large datasets.
4. **Performance**: Processing large amounts of data quickly and efficiently.

## Pandas: Your Data Analysis Swiss Army Knife

[Pandas](https://pandas.pydata.org) is a powerful, open-source library for Python that addresses these challenges and more. It provides high-performance, easy-to-use data structures and tools for data manipulation and analysis.

### Key Features of Pandas:

1. **DataFrame**: A two-dimensional labeled data structure with columns of potentially different types.
2. **Series**: A one-dimensional labeled array that can hold data of any type.
3. **Efficient data manipulation**: Tools for reading, writing, and transforming data.
4. **Handling of missing data**: Built-in support for handling missing data.
5. **Merging and joining datasets**: Combining multiple datasets easily.
6. **Time series functionality**: Tools for working with date and time data.
7. **Integration with other libraries**: Works well with [NumPy](https://numpy.org), [Matplotlib](https://matplotlib.org), and [scikit-learn](https://scikit-learn.org).

### The Role of Pandas in Data Analysis

Pandas serves as a crucial tool in the data analysis pipeline:

1. **Data Loading**: Pandas can read data from various file formats and databases.
2. **Data Cleaning**: It provides functions to handle missing values, remove duplicates, and format data.
3. **Data Transformation**: Pandas allows you to reshape, merge, and pivot your data.
4. **Data Analysis**: You can perform complex operations, grouping, and aggregations on your data.
5. **Data Visualisation**: While not a visualisation library itself, Pandas integrates well with plotting libraries like Matplotlib.

Let's explore these features hands-on and see how Pandas can make your data analysis tasks more efficient and enjoyable.

## Getting Started

Let's begin by importing Pandas and creating a simple `DataFrame` to explore its basic functionality. We'll use a dictionary to create a `DataFrame` with some sample data.

In [1]:
# Import pandas - by convention it is aliased as `pd`
import pandas as pd

# Create a simple DataFrame
data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 28],
    'City': ['New York', 'San Francisco', 'Los Angeles', 'Chicago']
}

df = pd.DataFrame(data)

# Display the DataFrame
df

Name  Age           City
0    Alice   25       New York
1      Bob   30  San Francisco
2  Charlie   35    Los Angeles
3    David   28        Chicago

Let's explore some basic operations that Pandas can perform on this `DataFrame`.

In [2]:
# Access a single column
ages = df['Age']

print(type(ages))

ages

<class 'pandas.core.series.Series'>


0    25
1    30
2    35
3    28
Name: Age, dtype: int64

In [3]:
# Access multiple columns
columns = df[['Name', 'City']]

print(type(columns))

columns

<class 'pandas.core.frame.DataFrame'>


Name           City
0    Alice       New York
1      Bob  San Francisco
2  Charlie    Los Angeles
3    David        Chicago

In [4]:
# Access a specific row by index
row = df.iloc[1]

print(type(row))

row

<class 'pandas.core.series.Series'>


Name              Bob
Age                30
City    San Francisco
Name: 1, dtype: object

In [5]:
# Access a specific cell
cell = df.loc[2, 'Age']

print(type(cell))

cell

<class 'numpy.int64'>


np.int64(35)

In [6]:
# Basic statistics
df['Age'].describe()

count     4.000000
mean     29.500000
std       4.203173
min      25.000000
25%      27.250000
50%      29.000000
75%      31.250000
max      35.000000
Name: Age, dtype: float64

In [7]:
# Filtering data
df[df['Age'] >= 30]

Name  Age           City
1      Bob   30  San Francisco
2  Charlie   35    Los Angeles

In [8]:
# Adding a new column
df['Country'] = 'USA'

df

Name  Age           City Country
0    Alice   25       New York     USA
1      Bob   30  San Francisco     USA
2  Charlie   35    Los Angeles     USA
3    David   28        Chicago     USA

In [9]:
# Simple data manipulation
df['Age in 5 years'] = df['Age'] + 5

df

Name  Age           City Country  Age in 5 years
0    Alice   25       New York     USA              30
1      Bob   30  San Francisco     USA              35
2  Charlie   35    Los Angeles     USA              40
3    David   28        Chicago     USA              33


These examples demonstrate some fundamental Pandas operations:

1. Creating a `DataFrame` from a dictionary
2. Accessing columns and rows
3. Retrieving basic statistics
4. Filtering data based on conditions
5. Adding new columns
6. Performing simple calculations on columns

As you can see, Pandas makes it easy to work with structured data. In the next sections, we'll explore more advanced features like data loading from files, data cleaning, and more complex manipulations.

## Data Loading

One of the most common tasks in data analysis is loading data from various sources. Pandas provides functions to read data from CSV files, Excel spreadsheets, SQL databases, JSON files, and more. For a complete list of supported file formats, refer to the Pandas [Input/output documentation](https://pandas.pydata.org/docs/reference/io.html).

Let's explore how to load data from a CSV file using Pandas. We'll use the `pd.read_csv()` function to read a CSV file into a `DataFrame`.

In [10]:
import os
import pandas as pd

# Titanic dataset - in this case, this will be loaded from a local folder (be aware of the file path i.e. any folders).
# data could also be loaded from Excel (spreadsheet) files, databases etc.
path ='titanic.csv'

titanic = pd.read_csv(path)

titanic

id  pclass  survived                                             name  \
0        1       1         1                    Allen, Miss. Elisabeth Walton   
1        2       1         1                   Allison, Master. Hudson Trevor   
2        3       1         0                     Allison, Miss. Helen Loraine   
3        4       1         0             Allison, Mr. Hudson Joshua Creighton   
4        5       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...    ...     ...       ...                                              ...   
1304  1305       3         0                             Zabour, Miss. Hileni   
1305  1306       3         0                            Zabour, Miss. Thamine   
1306  1307       3         0                        Zakarian, Mr. Mapriededer   
1307  1308       3         0                              Zakarian, Mr. Ortin   
1308  1309       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked  
0     female  29.0000      0      0   24160  211.3375       B5        S  
1       male   0.9167      1      2  113781  151.5500  C22 C26        S  
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  
...      ...      ...    ...    ...     ...       ...      ...      ...  
1304  female  14.5000      1      0    2665   14.4542      NaN        C  
1305  female      NaN      1      0    2665   14.4542      NaN        C  
1306    male  26.5000      0      0    2656    7.2250      NaN        C  
1307    male  27.0000      0      0    2670    7.2250      NaN        C  
1308    male  29.0000      0      0  315082    7.8750      NaN        S  

[1309 rows x 12 columns]

Once loaded, we can explore the data, check its structure, and perform various operations on it. This is the first step in the data analysis process: getting the data into a format that we can work with.

In [11]:
# The shape of the DataFrame - a tuple: (rows, columns)
titanic.shape

(1309, 12)

In [12]:
# Display a summary of the DataFrame
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        1309 non-null   int64  
 1   pclass    1309 non-null   int64  
 2   survived  1309 non-null   int64  
 3   name      1309 non-null   object 
 4   sex       1309 non-null   object 
 5   age       1046 non-null   float64
 6   sibsp     1309 non-null   int64  
 7   parch     1309 non-null   int64  
 8   ticket    1309 non-null   object 
 9   fare      1308 non-null   float64
 10  cabin     295 non-null    object 
 11  embarked  1307 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 122.8+ KB


In [13]:
# Display the first few rows of the DataFrame
titanic.head()

id  pclass  survived                                             name  \
0   1       1         1                    Allen, Miss. Elisabeth Walton   
1   2       1         1                   Allison, Master. Hudson Trevor   
2   3       1         0                     Allison, Miss. Helen Loraine   
3   4       1         0             Allison, Mr. Hudson Joshua Creighton   
4   5       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

      sex      age  sibsp  parch  ticket      fare    cabin embarked  
0  female  29.0000      0      0   24160  211.3375       B5        S  
1    male   0.9167      1      2  113781  151.5500  C22 C26        S  
2  female   2.0000      1      2  113781  151.5500  C22 C26        S  
3    male  30.0000      1      2  113781  151.5500  C22 C26        S  
4  female  25.0000      1      2  113781  151.5500  C22 C26        S

In [14]:
# Display the last few rows of the DataFrame
titanic.tail()

id  pclass  survived                       name     sex   age  sibsp  \
1304  1305       3         0       Zabour, Miss. Hileni  female  14.5      1   
1305  1306       3         0      Zabour, Miss. Thamine  female   NaN      1   
1306  1307       3         0  Zakarian, Mr. Mapriededer    male  26.5      0   
1307  1308       3         0        Zakarian, Mr. Ortin    male  27.0      0   
1308  1309       3         0         Zimmerman, Mr. Leo    male  29.0      0   

      parch  ticket     fare cabin embarked  
1304      0    2665  14.4542   NaN        C  
1305      0    2665  14.4542   NaN        C  
1306      0    2656   7.2250   NaN        C  
1307      0    2670   7.2250   NaN        C  
1308      0  315082   7.8750   NaN        S

We can use `groupby()` to group the data by a specific column and perform aggregations on it. This is a powerful feature that allows us to summarise and analyse data efficiently.

In [15]:
# Display a random sample of the DataFrame
titanic.sample(5)

id  pclass  survived  \
97      98       1         1   
1308  1309       3         0   
735    736       3         1   
272    273       1         1   
361    362       2         1   

                                                   name     sex   age  sibsp  \
97    Douglas, Mrs. Frederick Charles (Mary Helene B...  female  27.0      1   
1308                                 Zimmerman, Mr. Leo    male  29.0      0   
735      Coutts, Mrs. William (Winnie 'Minnie' Treanor)  female  36.0      0   
272       Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
361   Caldwell, Mrs. Albert Francis (Sylvia Mae Harb...  female  22.0      1   

      parch      ticket      fare    cabin embarked  
97        1    PC 17558  247.5208  B58 B60        C  
1308      0      315082    7.8750      NaN        S  
735       2  C.A. 37671   15.9000      NaN        S  
272       0       21228   82.2667      B45        S  
361       1      248738   29.0000      NaN        S

In [16]:
# Group by `pclass` and calculate the mean of the `fare` column
titanic.groupby(['pclass'])['fare'].mean()

pclass
1    87.508992
2    21.179196
3    13.302889
Name: fare, dtype: float64

In [17]:
# Group by `sex` and then `survived`, then count
survivals = titanic.groupby(['sex', 'survived']).size()

survivals

sex     survived
female  0           127
        1           339
male    0           682
        1           161
dtype: int64

## Data Cleaning

Real-world data is often messy and requires cleaning before analysis. This involves handling missing values, removing duplicates, and fixing inconsistencies in the data. Pandas provides functions to help with these tasks.

### Handling Missing Data

Missing data is a common issue in datasets and can affect the quality of our analysis. Pandas provides several functions to handle missing data, such as `isnull()`, `notnull()`, `dropna()`, and `fillna()`. These functions allow us to identify missing values, remove rows or columns with missing data, or fill in missing values with a specified value.

Let's explore how to handle missing data in a `DataFrame` using Pandas.

In [18]:
# Find the number of missing values in each column - `isna()` returns a DataFrame of booleans and `sum()` sums the columns
titanic.isna().sum()

id             0
pclass         0
survived       0
name           0
sex            0
age          263
sibsp          0
parch          0
ticket         0
fare           1
cabin       1014
embarked       2
dtype: int64

> Python will coerce a boolean value into an integer value when needed, so `True` becomes `1` and `False` becomes `0`. This is why we can use the `sum()` function to count the number of missing values (`isna() => True`) in each column.

#### Handling Missing Values - Cabin

There are 1,309 records in this dataset, and there are a significant number of missing values in the `cabin` column.

We can use the `dropna()` function to remove rows with missing values in the `cabin` column, but this would result in losing a large portion of the dataset. Another approach might be to drop the `cabin` column entirely, but we should consider the impact of this decision on our analysis. What if the `cabin` column contains valuable information that we need for our analysis?

Another option is to fill in the missing values with a placeholder value, such as `'None'`. This approach allows us to retain the information in the `cabin` column while handling the missing data.


In [19]:
# Replace missing values in the `cabin` column with the string 'None'
titanic['cabin'] = titanic['cabin'].fillna('None')

### Handling Missing Values - Age

The `age` column also has some missing values, but we can fill them in with the mean age of the passengers. This way, we retain the data while handling the missing values.

In [20]:
# Fill missing values in `age` column with the mean
titanic['age'] = titanic['age'].fillna(titanic['age'].mean())

titanic.sample(5)

id  pclass  survived  \
97      98       1         1   
1263  1264       3         0   
910    911       3         0   
349    350       2         1   
294    295       1         0   

                                                   name     sex   age  sibsp  \
97    Douglas, Mrs. Frederick Charles (Mary Helene B...  female  27.0      1   
1263                  van Billiard, Master. Walter John    male  11.5      1   
910                          Kallio, Mr. Nikolai Erland    male  17.0      0   
349                       Brown, Miss. Amelia 'Mildred'  female  24.0      0   
294                            Thayer, Mr. John Borland    male  49.0      1   

      parch             ticket      fare    cabin embarked  
97        1           PC 17558  247.5208  B58 B60        C  
1263      1           A/5. 851   14.5000     None        S  
910       0  STON/O 2. 3101274    7.1250     None        S  
349       0             248733   13.0000      F33        S  
294       1              17421  110.8833      C68        C

### Handling Missing Values - Embarked

The `embarked` column has only two missing values, which we can drop without losing much information, but it's also possible to fill them in with the most common value.

In [21]:
# Fill missing values in `embarked` column with the mode
titanic['embarked'] = titanic['embarked'].fillna(titanic['embarked'].mode()[0])

titanic.sample(5)

id  pclass  survived                            name     sex      age  \
590  591       2         1           West, Miss. Barbara J  female   0.9167   
211  212       1         0  Moore, Mr. Clarence Bloomfield    male  47.0000   
511  512       2         0       Myles, Mr. Thomas Francis    male  62.0000   
181  182       1         1     Leader, Dr. Alice (Farnham)  female  49.0000   
570  571       2         1             Toomey, Miss. Ellen  female  50.0000   

     sibsp  parch        ticket     fare cabin embarked  
590      1      2    C.A. 34651  27.7500  None        S  
211      0      0        113796  42.4000  None        S  
511      0      0        240276   9.6875  None        Q  
181      0      0         17465  25.9292   D17        S  
570      0      0  F.C.C. 13531  10.5000  None        S

### Handling Missing Values - Fare

The `fare` column only has one missing value, but we can't simply fill it with eith the mean or median value, as the fare is likely to be related to other factors like the passenger class or the port of embarkation.

Let's have a look at the row in question:

In [22]:
# Find rows where `fare` is missing
index = titanic[titanic['fare'].isna()].index

titanic.iloc[index]

id  pclass  survived                name   sex   age  sibsp  parch  \
1225  1226       3         0  Storey, Mr. Thomas  male  60.5      0      0   

     ticket  fare cabin embarked  
1225   3701   NaN  None        S

We can impute the median fare for the group of passengers that were in the same class and embarked from the same port, and use that value to fill in the missing fare value. We use median instead of mean as the fare distribution is unlikely to be continuous and may have outliers.

To get the median fare for each group, we can use the `groupby()` function to group the data by passenger class and port of embarkation, and then calculate the median fare for each group.

In [23]:
# Group by `pclass` and `embarked` and expose the `fare` column
fare_by_pclass_and_embarked = titanic.groupby(['pclass', 'embarked'])['fare']
fare_by_pclass_and_embarked.median()

pclass  embarked
1       C           76.7292
        Q           90.0000
        S           52.0000
2       C           15.3146
        Q           12.3500
        S           15.3750
3       C            7.8958
        Q            7.7500
        S            8.0500
Name: fare, dtype: float64

To get the fare, we need to index the resulting `fare_by_pclass_and_embarked` object by the passenger class and port of embarkation. The resulting code is a little difficult to read, but here it is:

In [24]:
# Get mean fare for 'pclass' = 3 and 'embarked' = 'S'
pclass = 3
embarked = 'S'

fare_by_pclass_and_embarked.median()[pclass][embarked]

np.float64(8.05)

We could do this manually, but we'll create a function to help automate the process.

In [25]:
def impute_fare(row):
    """
    Impute missing values in the `fare` column by taking the meadian of the fares of passengers with the same `pclass` and `embarked` values.

    Parameters:
    row: A row in the DataFrame

    Returns:
    The fare value if it is not missing, otherwise the median fare of passengers with the same `pclass` and `embarked` values
    """
    if pd.isna(row['fare']):
        pclass = row['pclass']
        embarked = row['embarked']
        return titanic.groupby(['pclass', 'embarked'])['fare'].median()[pclass][embarked]
    else:
        return row['fare']

# Apply the `impute_fare` function to the DataFrame
titanic['fare'] = titanic.apply(impute_fare, axis=1)

titanic.iloc[index] # Check that the record with the missing fare has been updated


id  pclass  survived                name   sex   age  sibsp  parch  \
1225  1226       3         0  Storey, Mr. Thomas  male  60.5      0      0   

     ticket  fare cabin embarked  
1225   3701  8.05  None        S

In [26]:
# Check that there are no missing values
assert titanic.isnull().sum().sum() == 0

Pandas provides other ways of handling missing data, such as [interpolation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html#pandas-dataframe-interpolate), [forward-fill](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ffill.html#pandas-dataframe-ffill), and [backward-fill](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.bfill.html#pandas-dataframe-bfill), depending on the context of the data. For more information, refer to the Pandas [missing data documentation](https://pandas.pydata.org/docs/reference/frame.html#missing-data-handling).

## Adding Features

In data analysis, we often need to create new features from existing data to extract more insights. Pandas makes it easy to add new columns to a `DataFrame` based on existing columns. We can perform calculations, transformations, or aggregations to create new features.

### Adding Features - Title

The `name` column in the Titanic dataset contains both the passenger's name and title. We can extract the title from the name and create a new column called `title` to store this information. This transformation can help us analyse the data based on the passenger's title.

> We're using a regular expression to extract the title from the name. The pattern `([A-Za-z]+)\.` captures one or more letters (`[A-Za-z]+`), followed by a period (`\.`). This pattern extracts the title from the name, such as 'Mr.', 'Mrs.', 'Miss', etc.

In [27]:
# Extract the title from the `name` column
titanic['title'] = titanic['name'].str.extract('([A-Za-z]+)\.', expand=False)
titanic['title'].value_counts()

title
Mr          757
Miss        260
Mrs         197
Master       61
Dr            8
Rev           8
Col           4
Major         2
Mlle          2
Ms            2
Lady          1
Capt          1
Mme           1
Sir           1
Jonkheer      1
Dona          1
Don           1
Countess      1
Name: count, dtype: int64

In [28]:
# Replace rare titles with 'Rare'
rare_titles = ['Dr', 'Rev', 'Col', 'Major', 'Lady', 'Capt', 'Sir', 'Jonkheer', 'Dona', 'Don', 'Countess']
titanic['title'] = titanic['title'].replace(rare_titles, 'Rare')

# Replace 'Mlle' and 'Ms' with 'Miss', and 'Mme' with 'Mrs'
titanic['title'] = titanic['title'].replace(['Mlle', 'Ms'], 'Miss')
titanic['title'] = titanic['title'].replace('Mme', 'Mrs')

titanic['title'].value_counts()

title
Mr        757
Miss      264
Mrs       198
Master     61
Rare       29
Name: count, dtype: int64

### Adding Features - One Hot Encoding

Categorical variables like `sex` and `embarked` need to be converted into numerical form for machine learning algorithms. One common technique is one-hot encoding, which creates binary columns for each category in the original column. Pandas provides a function called `get_dummies()` to perform one-hot encoding on categorical variables.

In [29]:
# One-hot encoding of `embarked` and `title` columns
categorical_columns = ['embarked', 'title']

titanic = pd.get_dummies(titanic, columns=categorical_columns, drop_first=True)

titanic.head()

id  pclass  survived                                             name  \
0   1       1         1                    Allen, Miss. Elisabeth Walton   
1   2       1         1                   Allison, Master. Hudson Trevor   
2   3       1         0                     Allison, Miss. Helen Loraine   
3   4       1         0             Allison, Mr. Hudson Joshua Creighton   
4   5       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

      sex      age  sibsp  parch  ticket      fare    cabin  embarked_Q  \
0  female  29.0000      0      0   24160  211.3375       B5       False   
1    male   0.9167      1      2  113781  151.5500  C22 C26       False   
2  female   2.0000      1      2  113781  151.5500  C22 C26       False   
3    male  30.0000      1      2  113781  151.5500  C22 C26       False   
4  female  25.0000      1      2  113781  151.5500  C22 C26       False   

   embarked_S  title_Miss  title_Mr  title_Mrs  title_Rare  
0        True        True     False      False       False  
1        True       False     False      False       False  
2        True        True     False      False       False  
3        True       False      True      False       False  
4        True       False     False       True       False

### Adding Features - Family Size and Alone

We can create a new feature called `family_size` by combining the `sibsp` (number of siblings/spouses aboard) and `parch` (number of parents/children aboard) columns. This new feature represents the total number of family members aboard with each passenger.

In [30]:
# Add a new column `family_size` by summing the `sibsp` and `parch` columns
titanic['family_size'] = titanic['sibsp'] + titanic['parch']

titanic.head()

id  pclass  survived                                             name  \
0   1       1         1                    Allen, Miss. Elisabeth Walton   
1   2       1         1                   Allison, Master. Hudson Trevor   
2   3       1         0                     Allison, Miss. Helen Loraine   
3   4       1         0             Allison, Mr. Hudson Joshua Creighton   
4   5       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

      sex      age  sibsp  parch  ticket      fare    cabin  embarked_Q  \
0  female  29.0000      0      0   24160  211.3375       B5       False   
1    male   0.9167      1      2  113781  151.5500  C22 C26       False   
2  female   2.0000      1      2  113781  151.5500  C22 C26       False   
3    male  30.0000      1      2  113781  151.5500  C22 C26       False   
4  female  25.0000      1      2  113781  151.5500  C22 C26       False   

   embarked_S  title_Miss  title_Mr  title_Mrs  title_Rare  family_size  
0        True        True     False      False       False            0  
1        True       False     False      False       False            3  
2        True        True     False      False       False            3  
3        True       False      True      False       False            3  
4        True       False     False       True       False            3

We can also create a new feature called `alone` to indicate whether a passenger was traveling alone or with family. This binary feature can help us analyse the survival rates of passengers traveling alone versus those traveling with family.

In [31]:
# Add a new column `alone` that is `True` if `family_size` is 0, otherwise `False`
titanic['alone'] = titanic['family_size'] == 0

titanic.head()

id  pclass  survived                                             name  \
0   1       1         1                    Allen, Miss. Elisabeth Walton   
1   2       1         1                   Allison, Master. Hudson Trevor   
2   3       1         0                     Allison, Miss. Helen Loraine   
3   4       1         0             Allison, Mr. Hudson Joshua Creighton   
4   5       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

      sex      age  sibsp  parch  ticket      fare    cabin  embarked_Q  \
0  female  29.0000      0      0   24160  211.3375       B5       False   
1    male   0.9167      1      2  113781  151.5500  C22 C26       False   
2  female   2.0000      1      2  113781  151.5500  C22 C26       False   
3    male  30.0000      1      2  113781  151.5500  C22 C26       False   
4  female  25.0000      1      2  113781  151.5500  C22 C26       False   

   embarked_S  title_Miss  title_Mr  title_Mrs  title_Rare  family_size  alone  
0        True        True     False      False       False            0   True  
1        True       False     False      False       False            3  False  
2        True        True     False      False       False            3  False  
3        True       False      True      False       False            3  False  
4        True       False     False       True       False            3  False

### Adding Features - Deck

We can use the `cabin` to determine the deck of the ship where the passenger's cabin was located. The deck information is encoded in the first character of the `cabin` value (e.g., `C85` corresponds to deck `C`). We can extract this information and create a new column called `deck` to store it.

- a `cabin` value of `'None'` will map to deck 0
- cabin `'A'` will map to deck 1, `'B'` to deck 2, and so on up to `'G'` (deck 7)
- cabins `'T'`, `'U'`, `'W'`, `'X'`, `'Y'`, and `'Z'` were on the boat deck, so we will map them to deck 8 (only one record has a cabin value of `'T'`)
- some records may contain multiple cabin values, separated by a space, in which case we will take the first cabin value to derive the deck

In [32]:
# Extract the deck from the `cabin` column
deck_mapping = {'N':0, 'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E': 5, 'F': 6, 'G': 7, 'T': 8, 'U': 8, 'W': 8, 'X': 8, 'Y': 8, 'Z': 8}

titanic['deck'] = titanic['cabin'].str[0].map(deck_mapping)

titanic.sample(5)

id  pclass  survived                         name   sex        age  \
1127  1128       3         0         Petroff, Mr. Nedelio  male  19.000000   
148    149       1         0  Harris, Mr. Henry Birkhardt  male  45.000000   
1284  1285       3         0            Webber, Mr. James  male  29.881135   
753    754       3         0             Davies, Mr. Evan  male  22.000000   
965    966       3         0       Lievens, Mr. Rene Aime  male  24.000000   

      sibsp  parch            ticket     fare cabin  embarked_Q  embarked_S  \
1127      0      0            349212   7.8958  None       False        True   
148       1      0             36973  83.4750   C83       False        True   
1284      0      0  SOTON/OQ 3101316   8.0500  None       False        True   
753       0      0       SC/A4 23568   8.0500  None       False        True   
965       0      0            345781   9.5000  None       False        True   

      title_Miss  title_Mr  title_Mrs  title_Rare  family_size  alone  deck  
1127       False      True      False       False            0   True     0  
148        False      True      False       False            1  False     3  
1284       False      True      False       False            0   True     0  
753        False      True      False       False            0   True     0  
965        False      True      False       False            0   True     0

## Transforming Features

Data transformation involves converting data into a suitable format for analysis. This may include scaling numerical features, encoding categorical variables, or normalising data. Pandas provides functions to help with these transformations.

### Transforming Features - Age

The `age` column contains continuous numerical data that can be normalised to improve the performance of our model. We can use the `StandardScaler` from `scikit-learn` to scale the `age` column to have a mean of 0 and a standard deviation of 1.

In [33]:
# Normalise the `age` column
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

titanic['age'] = scaler.fit_transform(titanic[['age']])

titanic.head()

id  pclass  survived                                             name  \
0   1       1         1                    Allen, Miss. Elisabeth Walton   
1   2       1         1                   Allison, Master. Hudson Trevor   
2   3       1         0                     Allison, Miss. Helen Loraine   
3   4       1         0             Allison, Mr. Hudson Joshua Creighton   
4   5       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

      sex       age  sibsp  parch  ticket      fare    cabin  embarked_Q  \
0  female -0.068420      0      0   24160  211.3375       B5       False   
1    male -2.249092      1      2  113781  151.5500  C22 C26       False   
2  female -2.164974      1      2  113781  151.5500  C22 C26       False   
3    male  0.009230      1      2  113781  151.5500  C22 C26       False   
4  female -0.379021      1      2  113781  151.5500  C22 C26       False   

   embarked_S  title_Miss  title_Mr  title_Mrs  title_Rare  family_size  \
0        True        True     False      False       False            0   
1        True       False     False      False       False            3   
2        True        True     False      False       False            3   
3        True       False      True      False       False            3   
4        True       False     False       True       False            3   

   alone  deck  
0   True     2  
1  False     3  
2  False     3  
3  False     3  
4  False     3

### Transforming Features - Sex

The `sex` column contains categorical data that can be converted into numerical form using one-hot encoding, but we can also map the values to integers directly.

In [34]:
# Map `sex` column to integer - male -> 0, female -> 1
titanic['sex'] = titanic['sex'].map({'male': 0, 'female': 1}).astype(int)

titanic.head()

id  pclass  survived                                             name  sex  \
0   1       1         1                    Allen, Miss. Elisabeth Walton    1   
1   2       1         1                   Allison, Master. Hudson Trevor    0   
2   3       1         0                     Allison, Miss. Helen Loraine    1   
3   4       1         0             Allison, Mr. Hudson Joshua Creighton    0   
4   5       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)    1   

        age  sibsp  parch  ticket      fare    cabin  embarked_Q  embarked_S  \
0 -0.068420      0      0   24160  211.3375       B5       False        True   
1 -2.249092      1      2  113781  151.5500  C22 C26       False        True   
2 -2.164974      1      2  113781  151.5500  C22 C26       False        True   
3  0.009230      1      2  113781  151.5500  C22 C26       False        True   
4 -0.379021      1      2  113781  151.5500  C22 C26       False        True   

   title_Miss  title_Mr  title_Mrs  title_Rare  family_size  alone  deck  
0        True     False      False       False            0   True     2  
1       False     False      False       False            3  False     3  
2        True     False      False       False            3  False     3  
3       False      True      False       False            3  False     3  
4       False     False       True       False            3  False     3

# Removing Unnecessary Columns

After creating new features, we may want to remove unnecessary columns from the dataset to reduce complexity and improve performance. Pandas provides the `drop()` function to remove columns from a `DataFrame`.

In [35]:
# Drop `cabin`, `name`, `parch`, and `sibsp` columns - no longer needed
titanic = titanic.drop(['cabin', 'name', 'parch', 'sibsp'], axis=1)

titanic.sample(5)

id  pclass  survived  sex           age    ticket      fare  \
779    780       3         0    1 -4.566709e-01    368702    7.7500   
926    927       3         1    0  2.758687e-16    368783    7.7500   
138    139       1         0    0  6.304310e-01  PC 17582  153.4625   
1018  1019       3         0    0  2.758687e-16    359306    8.0500   
450    451       2         0    0  1.562233e+00    250643   13.0000   

      embarked_Q  embarked_S  title_Miss  title_Mr  title_Mrs  title_Rare  \
779         True       False        True     False      False       False   
926         True       False       False      True      False       False   
138        False        True       False      True      False       False   
1018       False        True       False      True      False       False   
450        False        True       False      True      False       False   

      family_size  alone  deck  
779             0   True     0  
926             0   True     0  
138             1  False     3  
1018            0   True     0  
450             0   True     0

In [36]:
# Drop `id``, and `ticket` columns - not useful
titanic = titanic.drop(['id', 'ticket'], axis=1)

titanic.sample(5)

pclass  survived  sex       age      fare  embarked_Q  embarked_S  \
486        2         0    0 -0.456671   10.5000       False        True   
193        1         1    1 -1.155522  211.3375       False        True   
1210       3         0    0  0.785731   27.9000       False        True   
21         1         1    1  1.329282   52.5542       False        True   
561        2         0    0  2.261084   13.5000       False        True   

      title_Miss  title_Mr  title_Mrs  title_Rare  family_size  alone  deck  
486        False      True      False       False            0   True     0  
193         True     False      False       False            1  False     2  
1210       False      True      False       False            5  False     0  
21         False     False       True       False            2  False     4  
561        False      True      False       False            0   True     0

## Writing to a CSV file

Now that we have cleaned and preprocessed the data, we can write it to a CSV file using the `to_csv()` method. It's a good practice to save the cleaned data to a file so that we can reuse it later for analysis or modelling.

In [37]:
# Save the cleaned DataFrame to a new CSV file
titanic.to_csv('titanic_cleaned.csv', index=False)

We can also split the data into training and testing sets before writing them to a file. This allows us to save the training and testing data separately for machine learning tasks.

In [38]:
# Split the data into features and target
X = titanic.drop('survived', axis=1)
y = titanic['survived']

# Save the features and target to separate CSV files
X.to_csv('titanic_features.csv', index=False)
y.to_csv('titanic_target.csv', index=False)